New Rule: at least business support enabled

bksarthak opened the pull request:
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

No issue number applicable

New rule for checking if Business Support is enabled.

jongogogo commented:
This rule just check if there is a support api available, not if it is business level. Everyone should have access to support API. Not really sure what is the logic here to determine if it is business level.

jongogogo commented:
Good reference: https://github.com/awslabs/aws-config-rules/issues/151

bksarthak commented:
As DescribeCases API is only allowed when Business support is enabled, this rule tests if an error is thrown when you try to use that API call. If ClientError exception is thrown it tells that business support is not enabled on the account. I've also added another exception code where it will show noncompliant if the role doesn't have enough permission to use Support API

bksarthak commented:
```
'''
  #####################################
  ##           Gherkin               ##
  #####################################  
  Rule Name:
  BUSINESS_SUPPORT_ENABLED
  
  Description:
  Check whether the AWS Account has Business support plans enabled.
  
  Trigger:
  Frequency every 3hours
  
  Reports on:
  AWS::::Account
  
  Rule Parameters:
  None
  
  Scenarios:
  Scenario: 1
    Given: AWS Account has ability to use Support API's DescribeCases call with results returned without ClientError Exception
     Then: Return COMPLIANT
  Scenario: 2
    Given: AWS Support API errors out with AccessDenied exception while using the DescribeCases API call. This means the role use by Config is not allowed to use the Support API.
     Then: Return NON_COMPLIANT
  Scenario: 3
    Given: AWS Support API call errors out with SubscriptionRequired exception while using the DescribeCases API call. This error means Business Plan is not currently used for the account
     Then: Return NON_COMPLIANT
  Scenario: 4
    Given: AWS Support API call errors out with any other boto3.client exceptions
     Then: Return NOT_APPLICABLE
  
  '''
```

Adding the updated Gherkin for review (apologies I didn't know Gherkin was to be submitted first)

jongogogo commented:
> 
> 
> As DescribeCases API is only allowed when Business support is enabled, this rule tests if an error is thrown when you try to use that API call. If ClientError exception is thrown it tells that business support is not enabled on the account. I've also added another exception code where it will show noncompliant if the role doesn't have enough permission to use Support API

Cool. I found the reference: https://aws.amazon.com/premiumsupport/plans/
So it means that your rule checks for Support plan Business or above (Enterprise), which is a good rule. Can you please change the Gherkin+Name of the rule to reflect that.

For your scenarios:
1) Good
2) No need, this is part of the boiler plate code
3) Good
4) No need, this mostly part of the boiler plate code. I do suggest on your `except` in your else statement to just to another `raise`.

bksarthak commented:
Changed rule name to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED` on both Gherkin and Folder name.
Put a `raise` on the last `except` statement

Updated the Gherkin as follows:
```
#####################################
  ##           Gherkin               ##
  #####################################  
  Rule Name:
  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
  
  Description:
  Check whether the AWS Account has Business or above(Enterprise) support plan  is enabled
  
  Trigger:
  Frequency every 3hours
  
  Reports on:
  AWS::::Account
  
  Rule Parameters:
  None
  
  Scenarios:
  Scenario: 1
    Given: AWS Account has ability to use Support API's DescribeCases call with results returned without ClientError Exception
     Then: Return COMPLIANT
  Scenario: 2
    Given: AWS Support API call errors out with SubscriptionRequired exception while using the DescribeCases API call. This error means Business Plan is not currently used for the account
     Then: Return NON_COMPLIANT
```

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+# Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Please change the name of the file to BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py

bksarthak replied:
file name changed

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
+  
+  Description:
+  Check whether the AWS Account has Business or above(Enterprise) support plan  is enabled
```
Check whether the AWS Account is subscribed to the AWS Business Support Plan or above (i.e. Enterprise).

bksarthak replied:
changed as per suggestion

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+  Check whether the AWS Account has Business or above(Enterprise) support plan  is enabled
+  
+  Trigger:
+  Frequency every 3hours
```
Periodic

bksarthak replied:
Done

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+  ##           Gherkin               ##
+  #####################################  
+  Rule Name:
+  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
```
Tab it. Idem for the rest of the Gherkin. Example: https://github.com/awslabs/aws-config-rules/blob/master/python/ELASTICSEARCH_IN_VPC_ONLY/ELASTICSEARCH_IN_VPC_ONLY.py

bksarthak replied:
tabbed Gherkin as per given example

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    """Form the evaluation(s) to be return to Config Rules
```
Remove those explanations. It is only an help for the early developers.

bksarthak replied:
removed explanation under evaluate_compliance

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+    3 -- if None or an empty string, list or dict is returned, the Boilerplate code will put a "shadow" evaluation to feedback that the evaluation took place properly
+    """
+
+    ###############################
```
Remove comment

bksarthak replied:
removed comment

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+
+    try:
+        support_client.describe_cases()
+        annotate = 'Account has business support plan subscribed'
```
No annotation needed when it is compliant.

bksarthak replied:
removed

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py:
```diff
@@ -0,0 +1,415 @@
+        rule_evaluations.append(build_evaluation(account_id, 'COMPLIANT', event, annotation=annotate))
+    except ClientError as error:
+        if error.response['Error']['Code'] == 'SubscriptionRequiredException':
+            annotate = 'Account does not have business support plan.'
```
This AWS Account is not subscribed to the AWS business Support plan or above.

bksarthak replied:
done

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py:
```diff
@@ -0,0 +1,163 @@
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('business-support-enabled')
```
Change to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

bksarthak replied:
changed to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py:
```diff
@@ -0,0 +1,163 @@
+import sys
```
Change name to: BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py

bksarthak replied:
changed name of the file

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py:
```diff
@@ -0,0 +1,163 @@
+
+RULE = __import__('business-support-enabled')
+
+class SampleTest(unittest.TestCase):
```
Remove this Class. 

We need the tests for the rule though. 
`class Compliance_Test(unittest.TestCase):`

bksarthak replied:
remove SampleTest class and added Compliance_Test class with two test cases as per Gherkin

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0", 
+  "Parameters": {
+    "CodeKey": "business-support-enabled.zip", 
```
Change to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

bksarthak replied:
changed

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json:
```diff
@@ -0,0 +1,12 @@
+  "Parameters": {
+    "CodeKey": "business-support-enabled.zip", 
+    "SourceRuntime": "python3.6", 
+    "SourcePeriodic": "Three_Hours", 
```
`TwentyFour_Hours`

bksarthak replied:
changed to `TwentyFour_Hours`

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json:
```diff
@@ -0,0 +1,12 @@
+    "CodeKey": "business-support-enabled.zip", 
+    "SourceRuntime": "python3.6", 
+    "SourcePeriodic": "Three_Hours", 
+    "RuleName": "business-support-enabled", 
```
`BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

bksarthak replied:
changed

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:
```diff
@@ -0,0 +1,392 @@
+    rule_evaluations = []
+    support_client = boto3.client('support')
+    account_id = event['accountId']
+
```
remove extra line

bksarthak replied:
removed extra line

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py:
```diff
@@ -0,0 +1,160 @@
+    def test_scenario_2(self):
+        resp_expected = []
+        response = RULE.lambda_handler(build_lambda_scheduled_event(), {})
+        resp_expected.append(build_expected_response('COMPLIANT', '123456789012', DEFAULT_RESOURCE_TYPE))
```
You need to mock your support API call before that. Something like that to make an return:
`SUPPORT_CLIENT_MOCK.describe_cases = MagicMock(return_value={})`

bksarthak replied:
done

bksarthak commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py:
```diff
@@ -0,0 +1,160 @@
+    def test_scenario_1(self):
+        resp_expected = []
+        response = RULE.lambda_handler(build_lambda_scheduled_event(), {})
+        resp_expected.append(build_expected_response('NON_COMPLIANT', '0123456789012', DEFAULT_RESOURCE_TYPE, annotation='This AWS Account is not subscribed to the AWS business Support plan or above.'))
```
completed with `side_effect`

jongogogo commented:
Please fix the error. They are all relevant.

Check those changes to fix the boilerplate. https://github.com/awslabs/aws-config-rdk/pull/158/files

bksarthak commented:
> Please fix the errors. They are all relevant.
> 
> Check those changes to fix the boilerplate. https://github.com/awslabs/aws-config-rdk/pull/158/files

committed as per guidance

jongogogo commented:
************* Module BUSINESS_SUPPORT_OR_ABOVE_ENABLED
BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:150:0: E0001: invalid syntax (, line 150) (syntax-error)

bksarthak commented:
> ************* Module BUSINESS_SUPPORT_OR_ABOVE_ENABLED
> BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:150:0: E0001: invalid syntax (, line 150) (syntax-error)

the commenting was off around line 150. Accidental block comment of code instead of comments.

bksarthak commented:
no errors on the linting for `BUSINESS_SUPPORT_OR_ABOVE_ENABLED` rule

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:
```diff
@@ -0,0 +1,397 @@
+
+    try:
+        support_client.describe_cases()
+        rule_evaluations.append(build_evaluation(account_id, 'COMPLIANT', event))
```
do a direct `return build_evaluation(account_id, 'COMPLIANT', event)` 
remove the variable `rule_evaluations`

bksarthak replied:
done

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:
```diff
@@ -0,0 +1,397 @@
+    except ClientError as error:
+        if error.response['Error']['Code'] == 'SubscriptionRequiredException':
+            annotate = 'This AWS Account is not subscribed to the AWS business Support plan or above.'
+            rule_evaluations.append(build_evaluation(account_id, 'NON_COMPLIANT', event, annotation=annotate))
```
Same, direct return

bksarthak replied:
done

jongogogo commented on python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:
```diff
@@ -0,0 +1,397 @@
+        if error.response['Error']['Code'] == 'SubscriptionRequiredException':
+            annotate = 'This AWS Account is not subscribed to the AWS business Support plan or above.'
+            rule_evaluations.append(build_evaluation(account_id, 'NON_COMPLIANT', event, annotation=annotate))
+        else:
```
The `else` can then by removed. and the `raise` move 1 indentations upper.

bksarthak replied:
brought `raise` one level higher and removed the `else`
